llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -195,6 +195,32 @@ static void annotateNonNullAndDereferenceable(CallInst *CI, ArrayRef<unsigned> A
   }
 }
 
+static void mergeAttributes(CallInst *CI, AttributeList Attributes) {
+  // TODO: make this a member function on Instruction?
+  CI->setAttributes(
+      AttributeList::get(CI->getContext(), {CI->getAttributes(), Attributes}));
+}
+
+static MemTransferInst *
+createMemcpyFromStringCopy(IRBuilderBase &B, Value *Dst, MaybeAlign DstAlign,
+                           Value *Src, MaybeAlign SrcAlign, Value *Size) {
+  // A string never contains tags, so we can set no_preserve_tags.
+  return cast<MemTransferInst>(B.CreateMemCpy(
+      Dst, DstAlign, Src, SrcAlign, Size, PreserveCheriTags::Unnecessary));
+}
+
+static MemTransferInst *
+createMemcpyFromStringCopyWithAttrs(IRBuilderBase &B, AttributeList Attributes,
```

Comment:
Confused as to why this needs to exist; why would a string copy ever need to preserve tags?

Reply:
The existing calls to `NewCI->setAttributes(CI->getAttributes());` will remove the attributes set by CreateMemCpy(). I'll have a look if I can clean this up a bit, there might be some new attribute APIs since I originally wrote the code.